### PR TITLE
Correct and Smooth example

### PR DESCRIPTION
**Pull request by Kacper-Pietkun**
SAR implementation of correct and smooth model based on [DGL's implementation](https://github.com/dmlc/dgl/tree/master/examples/pytorch/correct_and_smooth)

Obtained results are similar to those presented in DGL:
image

This PR should be merged after #6

### CODE REVIEW

File: examples/correct_and_smooth.py [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+        global_features_sum = torch.mul(workers_means, workers_feature_sizes).sum(dim=0)
+        global_feature_size = workers_feature_sizes.sum()
+        global_means = global_features_sum / global_feature_size
+        
+        local_std_numerator = torch.pow(features - global_means, 2).sum(dim=0)
+        workers_std_numerators = sar.comm.exchange_tensors([local_std_numerator] * args.world_size)
+        workers_std_numerators = torch.stack(workers_std_numerators, dim=0)
+        global_stds = torch.sqrt(workers_std_numerators.sum(dim=0) / global_feature_size)
+        
+        features = (features - global_means) / global_stds
```
Comment (bgawrych) [inline examples/correct_and_smooth.py:410]: Can it be moved to a new function?

File: examples/correct_and_smooth.py [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
```
Comment (bgawrych) [inline examples/correct_and_smooth.py:19]: I'm not sure if it can be licensed this way, need to check

**Review by bgawrych — COMMENTED** [review]
Can you also add short description in README.md file

File: examples/correct_and_smooth.py (hunk 2 above)
Comment (Kacper-Pietkun) [reply examples/correct_and_smooth.py]: Ok, I deleted it since DGL repo is under Apache 2.0 and not MIT license. I mentioned link to original code in README.md and shortly described applied changes.

File: examples/correct_and_smooth.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+    # Sum the n_correct, and number of mask elements across all workers
+    sar.comm.all_reduce(acc_vec, op=dist.ReduceOp.SUM, move_to_comm_device=True)
+    (train_acc, val_acc, test_acc) =  \
+        (acc_vec[0] / acc_vec[1],
+         acc_vec[2] / acc_vec[3],
+         acc_vec[4] / acc_vec[5])
+    return train_acc, val_acc, test_acc
+
+
+def data_normalization(features, world_size):
```
Comment (hesham-mostafa) [inline examples/correct_and_smooth.py:338]: Can't this be done using our distributed BatchNorm layer that is included with SAR

Comment (hesham-mostafa) [reply examples/correct_and_smooth.py:338]: You can also use the VarOp and MeanOp from distributed_bn.py if you don't require the whole batchnorm layer.

Comment (Kacper-Pietkun) [reply examples/correct_and_smooth.py]: Ok, I've modified this part of code to use MeanOp and VarOp

File: examples/correct_and_smooth.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+            if self.autoscale:
+                smoothed_error = self.prop1(
+                    g, error, post_step=lambda x: x.clamp_(-1.0, 1.0)
+                )
+                sigma = error[mask].abs().sum() / numel
+                scale = sigma / smoothed_error.abs().sum(dim=1, keepdim=True)
+                scale[scale.isinf() | (scale > 1000)] = 1.0
+
+                result = y_soft + scale * smoothed_error
+                result[result.isnan()] = y_soft[result.isnan()]
```
Comment (bgawrych) [inline examples/correct_and_smooth.py:278]: I think in this fragment of code we need to synchronize this scale/sigma across all workers

Comment (hesham-mostafa) [reply examples/correct_and_smooth.py:278]: Strictly speaking. Yes. To do that, we have to sum the 'numel' across all clients,  sum error[mask].abs().sum() across all clients, then divide the two in line 273 to get sigma. We also need to take care of summing smoothed_error.abs().sum(..) across all clients.

Comment (Kacper-Pietkun) [reply examples/correct_and_smooth.py]: Ok, I've synchronized sigma across all workers. On the other hand, scale cannot be synchronized because `smoothed_error.abs().sum(dim=1, keepdim=True)` has different dimension for each worker (number of nodes in partition)

File: examples/README.md [hunk 5]
```diff
@@ -42,3 +42,15 @@ python /home/ubuntu/workspace/dgl/tools/launch.py \
     --ip_config ip_config.txt \
     "/home/ubuntu/miniconda3/bin/python train_distdgl_with_sar_inference.py --graph_name ogbn-products --ip_config ip_config.txt --num_epochs 2 --batch_size 1000 --part_config partition_data/ogbn-products.json"
 ```
+
+## Correct and Smooth
+Example taken from [DGL implemenetation](https://github.com/dmlc/dgl/tree/master/examples/pytorch/correct_and_smooth) of C&S. Code is adjusted to perform distributed training with SAR. For instance, you can run the example with following commands:
+
+* **Plain MLP + C&S**
+```shell
+python correct_and_smooth.py --partitioning-json-file /path/to/partitioning/graph_name.json --ip-file /path/to/ip_file --rank 0 --world-size 2 --dropout 0.5 --correction-adj DA --smoothing-adj AD --autoscale
```
Comment (bgawrych) [inline examples/README.md:51]: provide two commands, one with rank 0 and one with rank 1

File: examples/README.md [hunk 6]
```diff
@@ -42,3 +42,15 @@ python /home/ubuntu/workspace/dgl/tools/launch.py \
...
+```
+
+* **Plain Linear + C&S**
+```shell
+python correct_and_smooth.py --partitioning-json-file /path/to/partitioning/graph_name.json --ip-file /path/to/ip_file --rank 0 --world-size 2 --model linear --dropout 0.5 --epochs 1000 --correction-alpha 0.87 --smoothing-alpha 0.81 --correction-adj AD --autoscale
```
Comment (bgawrych) [inline examples/README.md:56]: same as above

File: examples/README.md [hunk 7]
```diff
@@ -42,3 +42,15 @@ python /home/ubuntu/workspace/dgl/tools/launch.py \
     --ip_config ip_config.txt \
     "/home/ubuntu/miniconda3/bin/python train_distdgl_with_sar_inference.py --graph_name ogbn-products --ip_config ip_config.txt --num_epochs 2 --batch_size 1000 --part_config partition_data/ogbn-products.json"
 ```
+
+## Correct and Smooth
+Example taken from [DGL implemenetation](https://github.com/dmlc/dgl/tree/master/examples/pytorch/correct_and_smooth) of C&S. Code is adjusted to perform distributed training with SAR. For instance, you can run the example with following commands:
```
Comment (bgawrych) [inline examples/README.md:47]: Add one sentence about changes to the original implementation: using DistributedBN1D and this scale synchronization

**Review by bgawrych — APPROVED** [review]
LGTM, minor comments

**Comment by Kacper-Pietkun** [comment]
@hesham-mostafa if you don't have any more comments, it is ready to merge
